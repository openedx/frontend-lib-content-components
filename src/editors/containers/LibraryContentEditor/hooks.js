import {
  useCallback,
  useEffect,
  useMemo,
  useState,
} from 'react';
import { useDispatch } from 'react-redux';

import { modes } from './constants';
import { actions } from '../../data/redux';
import * as requests from './data/requests';
import { RequestKeys, RequestStates } from '../../data/constants/requests';
import { isV1Library, getCandidates, getSelectedRows } from './utils';

export const blockTypeDisplay = (type) => {
  if (type === 'html') { return 'Text'; }
  if (type === 'video') { return 'Video'; }
  if (type === 'problem') { return 'Problem'; }
  return 'Other';
};

export const useLibraryHook = ({
  blockFailed,
  blockFinished,
  blockValue,
}) => {
  const dispatch = useDispatch();

  // load previously saved block data from blockValue
  useEffect(() => {
    const metadata = blockValue?.data?.metadata;
    const libraryId = metadata?.source_library_id ?? null;
    let settings = {};
    if (libraryId) {
      settings = {
        [libraryId]: {
          version: metadata?.source_library_version,
          mode: metadata?.manual ? modes.selected.value : modes.random.value,
          count: metadata?.max_count,
          showReset: metadata?.allow_resetting_children,
          candidates: metadata?.candidates,
          blocks: [],
        },
      };
    }
    dispatch(actions.library.initializeFromBlockValue({ libraryId, settings }));
  }, [blockValue]);

  // fetch libraries when block finishes loading
  useEffect(() => {
    if (blockFinished && !blockFailed) {
      dispatch(requests.fetchV2Libraries({
        onSuccess: (response) => {
          const libraries = {};
          response?.data.forEach(library => {
            libraries[library.id] = library;
          });
          dispatch(actions.library.addLibraries({ libraries }));
        },
        onFailure: (error) => {
          dispatch(actions.requests.failRequest({
            requestKey: RequestKeys.fetchV2Libraries,
            error,
          }));
        },
      }));
      dispatch(requests.fetchV1Libraries({
        onSuccess: (response) => {
          const libraries = {};
          response?.data?.libraries.forEach(library => {
            libraries[library.library_key] = library;
          });
          dispatch(actions.library.addLibraries({ libraries }));
        },
        onFailure: (error) => {
          dispatch(actions.requests.failRequest({
            requestKey: RequestKeys.fetchV1Libraries,
            error,
          }));
        },
      }));
    }
  }, [blockFinished, blockFailed]);
};

export const useLibrarySelectorHook = ({
  selectedLibraryId,
  settings,
}) => {
  const dispatch = useDispatch();

  return {
    onLibrarySelect: (id) => {
      if (id !== selectedLibraryId) {
        dispatch(actions.library.setLibraryId({ selectedLibraryId: id }));
        if (!settings[id]) {
          dispatch(actions.library.initialLibrarySettings({ selectedLibraryId: id }));
        }
      }
    },
  };
};

export const useBlocksSelectorHook = ({
  blocks,
  candidates,
  libraries,
  savedLibraryId,
  selectedLibraryId,
  v1BlockRequests,
}) => {
  const dispatch = useDispatch();

  const [tableDataLoaded, setTableDataLoaded] = useState(false);

  // fetch library version and blocks
  useEffect(() => {
    setTableDataLoaded(false);
    if (!blocks || blocks.length === 0) {
      if (selectedLibraryId === savedLibraryId) {
        dispatch(requests.fetchChildrenInfo({
          onSuccess: (response) => {
            const children = response?.data?.children;
            dispatch(actions.library.setLibraryBlocks({
              blocks: children.map(block => ({
                id: block.id,
                display_name: block.display_name,
                block_type: block.category,
              })),
            }));
          },
          onFailure: (error) => {
            dispatch(actions.requests.failRequest({
              requestKey: RequestKeys.fetchChildrenInfo,
              error,
            }));
          },
        }));
      } else if (isV1Library(selectedLibraryId)) {
        if (!v1BlockRequests || Object.keys(v1BlockRequests).length === 0) {
          dispatch(requests.fetchV1LibraryContent({
            libraryId: selectedLibraryId,
            onSuccess: (response) => {
              const v1Blocks = response.data?.blocks ?? [];
              const initialRequestStatus = {};
              v1Blocks.forEach(id => {
                initialRequestStatus[id] = RequestStates.inactive;
              });
              dispatch(actions.library.setLibraryVersion({
                version: response.data?.version,
              }));
              dispatch(actions.library.setV1BlockRequests({
                v1BlockRequests: initialRequestStatus,
              }));
            },
            onFailure: (error) => {
              dispatch(actions.requests.failRequest({
                requestKey: RequestKeys.fetchV1LibraryContent,
                error,
              }));
            },
          }));
        }
      } else {
        dispatch(actions.library.setLibraryVersion({
          version: libraries[selectedLibraryId]?.version,
        }));
        dispatch(requests.fetchV2LibraryContent({
          libraryId: selectedLibraryId,
          onSuccess: (response) => {
            dispatch(actions.library.setLibraryBlocks({
              blocks: response?.data,
            }));
          },
          onFailure: (error) => {
            dispatch(actions.requests.failRequest({
              requestKey: RequestKeys.fetchV2LibraryContent,
              error,
            }));
          },
        }));
      }
    }
  }, [selectedLibraryId]);

  // set up blocks for v1 library
  useEffect(() => {
    if (blocks.length < Object.keys(v1BlockRequests).length) {
      if (isV1Library(selectedLibraryId)) {
        Object.keys(v1BlockRequests).forEach(blockId => {
          const status = v1BlockRequests[blockId];
          if (status === RequestStates.inactive) {
            dispatch(actions.library.updateV1BlockRequestStatus({
              blockId,
              status: RequestStates.pending,
            }));
            dispatch(requests.fetchV1LibraryBlock({
              blockId,
              onSuccess: (response) => {
                const v1Block = {
                  id: blockId,
                  display_name: response.data?.display_name ?? '',
                  block_type: response.data?.category ?? '',
                };
                dispatch(actions.library.addLibraryBlock({ block: v1Block }));
                dispatch(actions.library.updateV1BlockRequestStatus({
                  blockId,
                  status: RequestStates.completed,
                }));
              },
              onFailure: () => {
                dispatch(actions.library.updateV1BlockRequestStatus({
                  blockId,
                  status: RequestStates.failed,
                }));
              },
            }));
          }
        });
      }
    }
  }, [v1BlockRequests]);

  // check if blocks for table is loaded
  useEffect(() => {
    let loaded = true;
    Object.keys(v1BlockRequests).forEach(blockId => {
      const status = v1BlockRequests[blockId];
      if ((status === RequestStates.inactive) || (status === RequestStates.pending)) {
        loaded = false;
      }
    });
    setTableDataLoaded(loaded);
  }, [v1BlockRequests]);

  return {
    tableDataLoaded,
    data: useMemo(() => {
      if (tableDataLoaded) {
        return blocks.map(block => ({
          display_name: block.display_name,
          block_type: blockTypeDisplay(block.block_type),
        }));
      }
      return [];
    }, [blocks, tableDataLoaded]),

    initialRows: useMemo(() => {
      if (tableDataLoaded) {
        return getSelectedRows({ blocks, candidates });
      }
      return {};
    }, [blocks, candidates, tableDataLoaded]),

    onSelectedRowsChanged: useCallback((selected) => {
      if (tableDataLoaded) {
        dispatch(actions.library.setCandidatesForLibrary({
          candidates: getCandidates({
            blocks,
            rows: selected,
          }),
        }));
      }
    }, [blocks, tableDataLoaded]),
  };
};
