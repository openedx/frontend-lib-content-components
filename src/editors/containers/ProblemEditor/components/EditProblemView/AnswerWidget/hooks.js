import { useState, useEffect } from 'react';
import { StrictDict } from '../../../../../utils';
import * as module from './hooks';
import { actions } from '../../../../../data/redux';
import { ProblemTypeKeys } from '../../../../../data/constants/problem';

export const state = StrictDict({
  isFeedbackVisible: (val) => useState(val),
});

export const removeAnswer = ({ answer, dispatch }) => () => {
  dispatch(actions.problem.deleteAnswer({ id: answer.id }));
};

export const setAnswer = ({ answer, hasSingleAnswer, dispatch }) => (payload) => {
  dispatch(actions.problem.updateAnswer({ id: answer.id, hasSingleAnswer, ...payload }));
};

export const prepareFeedback = (answer) => {
  const [isFeedbackVisible, setIsFeedbackVisible] = module.state.isFeedbackVisible(false);
  useEffect(() => {
    // Show feedback fields if feedback is present
    const isVisible = !!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback;
    setIsFeedbackVisible(isVisible);
  }, [answer]);

  const toggleFeedback = (open) => {
    // Do not allow to hide if feedback is added
    if (!!answer.selectedFeedback || !!answer.unselectedFeedback || !!answer.feedback) {
      setIsFeedbackVisible(true);
      return;
    }
    setIsFeedbackVisible(open);
  };
  return {
    isFeedbackVisible,
    toggleFeedback,
  };
};

export const isSingleAnswerProblem = (problemType) => (
  problemType === ProblemTypeKeys.DROPDOWN || problemType === ProblemTypeKeys.SINGLESELECT
);

export const useAnswerContainer = ({ answers, updateField }) => {
  useEffect(() => {
    let answerCount = 0;
    answers.forEach(answer => {
      if (answer.correct) {
        answerCount += 1;
      }
    });
    updateField({ correctAnswerCount: answerCount });
  }, []);
};

export default {
  state, removeAnswer, setAnswer, prepareFeedback, isSingleAnswerProblem, useAnswerContainer,
};
