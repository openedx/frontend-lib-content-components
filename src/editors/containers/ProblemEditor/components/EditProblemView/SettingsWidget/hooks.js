import { useState, useEffect } from 'react';

import _ from 'lodash-es';
import * as module from './hooks';
import messages from './messages';
import { ProblemTypeKeys, ShowAnswerTypesKeys } from '../../../../../data/constants/problem';

export const state = {
  showAdvanced: (val) => useState(val),
  cardCollapsed: (val) => useState(val),
  summary: (val) => useState(val),
  showAttempts: (val) => useState(val),
};

export const showAdvancedSettingsCards = () => {
  const [isAdvancedCardsVisible, setIsAdvancedCardsVisible] = module.state.showAdvanced(false);
  return {
    isAdvancedCardsVisible,
    showAdvancedCards: () => setIsAdvancedCardsVisible(true),
  };
};

export const showFullCard = () => {
  const [isCardCollapsed, setIsCardCollapsed] = module.state.cardCollapsed(false);
  return {
    isCardCollapsed,
    toggleCardCollapse: () => setIsCardCollapsed(!isCardCollapsed),
  };
};

export const hintsCardHooks = (hints, updateSettings) => {
  const [summary, setSummary] = module.state.summary({ message: messages.noHintSummary, values: {} });

  useEffect(() => {
    const hintsNumber = hints.length;
    if (hintsNumber === 0) {
      setSummary({ message: messages.noHintSummary, values: {} });
    } else {
      setSummary({ message: messages.hintSummary, values: { hint: hints[0].value, count: (hintsNumber - 1) } });
    }
  }, [hints]);

  const handleAdd = () => {
    let newId = 0;
    if (!_.isEmpty(hints)) {
      newId = Math.max(...hints.map(hint => hint.id)) + 1;
    }
    const hint = { id: newId, value: '' };
    const modifiedHints = [...hints, hint];
    updateSettings({ hints: modifiedHints });
  };

  return {
    summary,
    handleAdd,
  };
};

export const hintsRowHooks = (id, hints, updateSettings) => {
  const handleChange = (event) => {
    const { value } = event.target;
    const modifiedHints = hints.map(hint => {
      if (hint.id === id) {
        return { ...hint, value };
      }
      return hint;
    });
    updateSettings({ hints: modifiedHints });
  };

  const handleDelete = () => {
    const modifiedHints = hints.filter((hint) => (hint.id !== id));
    updateSettings({ hints: modifiedHints });
  };

  return {
    handleChange,
    handleDelete,
  };
};

export const matlabCardHooks = (matLabApiKey, updateSettings) => {
  const [summary, setSummary] = module.state.summary({ message: '', values: {}, intl: false });

  useEffect(() => {
    if (_.isEmpty(matLabApiKey)) {
      setSummary({ message: messages.matlabNoKeySummary, values: {}, intl: true });
    } else {
      setSummary({ message: matLabApiKey, values: {}, intl: false });
    }
  }, [matLabApiKey]);

  const handleChange = (event) => {
    updateSettings({ matLabApiKey: event.target.value });
  };

  return {
    summary,
    handleChange,
  };
};

export const resetCardHooks = (updateSettings) => {
  const setReset = (value) => {
    updateSettings({ showResetButton: value });
  };

  return {
    setResetTrue: () => setReset(true),
    setResetFalse: () => setReset(false),
  };
};

export const scoringCardHooks = (scoring, updateSettings) => {
  const handleMaxAttemptChange = (event) => {
    let unlimitedAttempts = false;
    let attemptNumber = parseInt(event.target.value);
    if (_.isNaN(attemptNumber)) {
      attemptNumber = null;
      unlimitedAttempts = true;
    } else if (attemptNumber < 0) {
      attemptNumber = 0;
    }
    updateSettings({ scoring: { ...scoring, attempts: { number: attemptNumber, unlimited: unlimitedAttempts } } });
  };

  const handleWeightChange = (event) => {
    let weight = parseFloat(event.target.value);
    if (_.isNaN(weight)) {
      weight = 0;
    }
    updateSettings({ scoring: { ...scoring, weight } });
  };

  return {
    handleMaxAttemptChange,
    handleWeightChange,
  };
};

export const showAnswerCardHooks = (showAnswer, updateSettings) => {
  const [showAttempts, setShowAttempts] = module.state.showAttempts(false);
  const numberOfAttemptsChoice = [
    ShowAnswerTypesKeys.AFTER_SOME_NUMBER_OF_ATTEMPTS,
    ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS,
    ShowAnswerTypesKeys.AFTER_ALL_ATTEMPTS_OR_CORRECT,
  ];

  useEffect(() => {
    setShowAttempts(_.includes(numberOfAttemptsChoice, showAnswer.on));
  }, [showAttempts]);

  const handleShowAnswerChange = (event) => {
    const { value } = event.target;
    setShowAttempts(_.includes(numberOfAttemptsChoice, value));
    updateSettings({ showAnswer: { ...showAnswer, on: value } });
  };

  const handleAttemptsChange = (event) => {
    let attempts = parseInt(event.target.value);
    if (_.isNaN(attempts)) {
      attempts = 0;
    }
    updateSettings({ showAnswer: { ...showAnswer, afterAttempts: attempts } });
  };

  return {
    handleShowAnswerChange,
    handleAttemptsChange,
    showAttempts,
  };
};

export const timerCardHooks = (updateSettings) => ({
  handleChange: (event) => {
    let time = parseInt(event.target.value);
    if (_.isNaN(time)) {
      time = 0;
    }
    updateSettings({ timeBetween: time });
  },
});

export const typeRowHooks = ({
  answers,
  correctAnswerCount,
  typeKey,
  updateField,
  updateAnswer,
}) => {
  const onClick = () => {
    if (typeKey === ProblemTypeKeys.SINGLESELECT || typeKey === ProblemTypeKeys.DROPDOWN) {
      if (correctAnswerCount > 1) {
        answers.forEach(answer => {
          updateAnswer({ ...answer, correct: false });
        });
      }
    }
    updateField({ problemType: typeKey });
  };
  return {
    onClick,
  };
};

export const confirmSwitchToAdvancedEditor = ({
  switchToAdvancedEditor,
  setConfirmOpen,
}) => {
  switchToAdvancedEditor();
  setConfirmOpen(false);
  window.scrollTo({
    top: 0,
    behavior: 'smooth',
  });
};
