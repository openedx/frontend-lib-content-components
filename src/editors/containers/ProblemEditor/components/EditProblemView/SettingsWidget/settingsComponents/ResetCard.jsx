import React from 'react';
import { FormattedMessage, injectIntl, intlShape } from '@edx/frontend-platform/i18n';
import { Button, ButtonGroup, Hyperlink } from '@edx/paragon';
import PropTypes from 'prop-types';
import SettingsOption from '../SettingsOption';
import messages from '../messages';
import { resetCardHooks } from '../hooks';

export const ResetCard = ({
  showResetButton,
  updateSettings,
  // inject
  intl,
}) => {
  const { setResetTrue, setResetFalse } = resetCardHooks(updateSettings);
  return (
    <SettingsOption
      title={intl.formatMessage(messages.resetSettingsTitle)}
      summary={showResetButton
        ? intl.formatMessage(messages.resetSettingsTrue) : intl.formatMessage(messages.resetSettingsFalse)}
      className="resetCard"
    >
      <div className="halfSpacedMessage">
        <span>
          <FormattedMessage {...messages.resetSettingText} />
        </span>
      </div>
      <div className="spacedMessage">
        <Hyperlink destination="#" target="_blank">
          <FormattedMessage {...messages.advancedSettingsLinkText} />
        </Hyperlink>
      </div>
      <ButtonGroup size="sm" className="resetSettingsButtons mb-2">
        <Button variant={showResetButton ? 'outline-primary' : 'primary'} size="sm" onClick={setResetFalse}>
          <FormattedMessage {...messages.resetSettingsFalse} />
        </Button>
        <Button variant={showResetButton ? 'primary' : 'outline-primary'} size="sm" onClick={setResetTrue}>
          <FormattedMessage {...messages.resetSettingsTrue} />
        </Button>
      </ButtonGroup>
    </SettingsOption>
  );
};

ResetCard.propTypes = {
  showResetButton: PropTypes.bool.isRequired,
  updateSettings: PropTypes.func.isRequired,
  // injected
  intl: intlShape.isRequired,
};

export default injectIntl(ResetCard);
