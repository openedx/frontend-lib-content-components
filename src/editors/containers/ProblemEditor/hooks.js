import {
  useRef, useCallback, useState, useEffect,
} from 'react';
import tinyMCEStyles from '../../data/constants/tinyMCEStyles';
import { StrictDict } from '../../utils';
import pluginConfig from '../TextEditor/pluginConfig';
import * as module from './hooks';

export const state = StrictDict({
  refReady: (val) => useState(val),
});

export const parseContentForLabels = ({ editor, updateQuestion }) => {
  let content = editor.getContent();
  const parsedLabels = content.split(/<label>|<\/label>/gm);
  let updatedContent;
  parsedLabels.forEach((label, i) => {
    let updatedLabel = label;
    if (!label.startsWith('<') && !label.endsWith('>')) {
      let previousLabel = parsedLabels[i - 1];
      let nextLabel = parsedLabels[i + 1];
      if (!previousLabel.endsWith('<p>')) {
        previousLabel = `${previousLabel}</p><p>`;
        updatedContent = content.replace(parsedLabels[i - 1], previousLabel);
        content = updatedContent;
      }
      if (previousLabel.endsWith('</p>') && !label.startWith('<p>')) {
        updatedLabel = `<p>${label}`;
        updatedContent = content.replace(label, updatedLabel);
        content = updatedContent;
      }
      if (!nextLabel.startsWith('</p>')) {
        nextLabel = `</p><p>${nextLabel}`;
        updatedContent = content.replace(parsedLabels[i + 1], nextLabel);
        content = updatedContent;
      }
    }
  });
  updateQuestion(content);
};

export const setupCustomBehavior = ({ updateQuestion }) => (editor) => {
  // add a custom simple inline label formatter.
  const toggleLabelFormatting = () => {
    editor.execCommand('mceToggleFormat', false, 'label');
  };
  editor.ui.registry.addIcon('textToSpeech',
    '<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg"><path d="M3.5 22C3.08333 22 2.72917 21.8542 2.4375 21.5625C2.14583 21.2708 2 20.9167 2 20.5V3.5C2 3.08333 2.14583 2.72917 2.4375 2.4375C2.72917 2.14583 3.08333 2 3.5 2H13L11.5 3.5H3.5V20.5H15.5V17H17V20.5C17 20.9167 16.8542 21.2708 16.5625 21.5625C16.2708 21.8542 15.9167 22 15.5 22H3.5ZM6 17.75V16.25H13V17.75H6ZM6 14.75V13.25H11V14.75H6ZM15.5 15L11.5 11H8V6H11.5L15.5 2V15ZM17 12.7V4.05C17.9333 4.4 18.6667 5.01667 19.2 5.9C19.7333 6.78333 20 7.65 20 8.5C20 9.35 19.7083 10.1917 19.125 11.025C18.5417 11.8583 17.8333 12.4167 17 12.7ZM17 16.25V14.7C18.1667 14.2833 19.2083 13.5333 20.125 12.45C21.0417 11.3667 21.5 10.05 21.5 8.5C21.5 6.95 21.0417 5.63333 20.125 4.55C19.2083 3.46667 18.1667 2.71667 17 2.3V0.75C18.7 1.2 20.125 2.1375 21.275 3.5625C22.425 4.9875 23 6.63333 23 8.5C23 10.3667 22.425 12.0125 21.275 13.4375C20.125 14.8625 18.7 15.8 17 16.25Z" fill="black"/></svg>');
  editor.ui.registry.addButton('customLabelButton', {
    icon: 'textToSpeech',
    text: 'Label',
    tooltip: 'Apply a "Question" label to specific text, recognized by screen readers. Recommended to improve accessibility.',
    onAction: toggleLabelFormatting,
  });
  editor.on('blur', () => {
    module.parseContentForLabels({
      editor,
      updateQuestion,
    });
  });
};

export const problemEditorConfig = ({
  setEditorRef,
  question,
  updateQuestion,
}) => ({
  onInit: (evt, editor) => {
    setEditorRef(editor);
  },
  initialValue: question || '',
  init: {
    skin: false,
    content_css: false,
    content_style: tinyMCEStyles,
    menubar: false,
    branding: false,
    min_height: 150,
    placeholder: 'Enter your question',
    formats: { label: { inline: 'label' } },
    setup: module.setupCustomBehavior({ updateQuestion }),
    toolbar: `${pluginConfig().toolbar} | customLabelButton`,
    plugins: 'autoresize',
  },
});

export const prepareEditorRef = () => {
  const editorRef = useRef(null);
  const setEditorRef = useCallback((ref) => {
    editorRef.current = ref;
  }, []);
  const [refReady, setRefReady] = module.state.refReady(false);
  useEffect(() => setRefReady(true), [setRefReady]);
  return { editorRef, refReady, setEditorRef };
};
