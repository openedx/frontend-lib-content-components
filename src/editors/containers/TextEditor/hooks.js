import * as appHooks from '../../hooks';
import { setAssetToStaticUrl } from '../../sharedComponents/TinyMceWidget/hooks';

export const { nullMethod, navigateCallback, navigateTo } = appHooks;

export const getContent = ({ editorRef, isRawEditor }) => () => {
  const content = (isRawEditor && editorRef && editorRef.current
    ? editorRef.current.state.doc.toString()
    : editorRef.current?.getContent());
  return setAssetToStaticUrl({ editorValue: content });
};
