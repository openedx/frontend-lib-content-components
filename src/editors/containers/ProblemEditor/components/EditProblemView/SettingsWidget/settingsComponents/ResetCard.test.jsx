import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import { formatMessage } from '../../../../../../../testUtils';
import { ResetCard } from './ResetCard';
import { resetCardHooks } from '../hooks';

jest.mock('../hooks', () => ({
  resetCardHooks: jest.fn(),
}));

jest.mock('../../../../../../data/redux', () => ({
  selectors: {
    app: {
      studioEndpointUrl: 'sTuDioEndpOintUrl',
      learningContextId: 'leArningCoNteXtId',
    },
  },
}));

jest.mock('react-redux', () => ({
  useSelector: jest.fn().mockImplementation((args) => args),
}));

describe('ResetCard', () => {
  const props = {
    showResetButton: false,
    updateSettings: jest.fn().mockName('args.updateSettings'),
    intl: { formatMessage },
  };

  const resetCardHooksProps = {
    setResetTrue: jest.fn().mockName('resetCardHooks.setResetTrue'),
    setResetFalse: jest.fn().mockName('resetCardHooks.setResetFalse'),
  };

  resetCardHooks.mockReturnValue(resetCardHooksProps);

  describe('behavior', () => {
    it(' calls resetCardHooks when initialized', () => {
      shallow(<ResetCard {...props} />);
      expect(resetCardHooks).toHaveBeenCalledWith(props.updateSettings);
    });
  });

  describe('snapshot', () => {
    test('snapshot: renders reset true setting card', () => {
      expect(shallow(<ResetCard {...props} />).snapshot).toMatchSnapshot();
    });
    test('snapshot: renders reset true setting card', () => {
      expect(shallow(<ResetCard {...props} showResetButton />).snapshot).toMatchSnapshot();
    });
  });
});
