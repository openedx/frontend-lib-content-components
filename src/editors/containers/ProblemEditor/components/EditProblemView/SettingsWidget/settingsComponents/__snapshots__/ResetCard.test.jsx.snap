// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResetCard snapshot snapshot: renders reset true setting card 1`] = `
<SettingsOption
  className="resetCard"
  none={false}
  summary="False"
  title="Show reset option"
>
  <div
    className="halfSpacedMessage"
  >
    <span>
      <FormattedMessage
        defaultMessage="Determines whether a 'Reset' button is shown so the user may reset their answer, generally for use in practice or formative assessments."
        description="Reset settings card text"
        id="authoring.problemeditor.settings.reset.text"
      />
    </span>
  </div>
  <div
    className="spacedMessage"
  >
    <Hyperlink
      destination="#"
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Set a default value in advanced settings"
        description="Advanced settings link text"
        id="authoring.problemeditor.settings.advancedSettingLink.text"
      />
    </Hyperlink>
  </div>
  <ButtonGroup
    className="resetSettingsButtons mb-2"
    size="sm"
  >
    <Button
      onClick={[MockFunction resetCardHooks.setResetFalse]}
      size="sm"
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="False"
        description="False option for reset"
        id="authoring.problemeditor.settings.reset.false"
      />
    </Button>
    <Button
      onClick={[MockFunction resetCardHooks.setResetTrue]}
      size="sm"
      variant="outline-primary"
    >
      <FormattedMessage
        defaultMessage="True"
        description="True option for reset"
        id="authoring.problemeditor.settings.reset.true"
      />
    </Button>
  </ButtonGroup>
</SettingsOption>
`;

exports[`ResetCard snapshot snapshot: renders reset true setting card 2`] = `
<SettingsOption
  className="resetCard"
  none={false}
  summary="True"
  title="Show reset option"
>
  <div
    className="halfSpacedMessage"
  >
    <span>
      <FormattedMessage
        defaultMessage="Determines whether a 'Reset' button is shown so the user may reset their answer, generally for use in practice or formative assessments."
        description="Reset settings card text"
        id="authoring.problemeditor.settings.reset.text"
      />
    </span>
  </div>
  <div
    className="spacedMessage"
  >
    <Hyperlink
      destination="#"
      target="_blank"
    >
      <FormattedMessage
        defaultMessage="Set a default value in advanced settings"
        description="Advanced settings link text"
        id="authoring.problemeditor.settings.advancedSettingLink.text"
      />
    </Hyperlink>
  </div>
  <ButtonGroup
    className="resetSettingsButtons mb-2"
    size="sm"
  >
    <Button
      onClick={[MockFunction resetCardHooks.setResetFalse]}
      size="sm"
      variant="outline-primary"
    >
      <FormattedMessage
        defaultMessage="False"
        description="False option for reset"
        id="authoring.problemeditor.settings.reset.false"
      />
    </Button>
    <Button
      onClick={[MockFunction resetCardHooks.setResetTrue]}
      size="sm"
      variant="primary"
    >
      <FormattedMessage
        defaultMessage="True"
        description="True option for reset"
        id="authoring.problemeditor.settings.reset.true"
      />
    </Button>
  </ButtonGroup>
</SettingsOption>
`;
